### PR TITLE
Optimize and add new efficient visual tree extensions.

### PR DESCRIPTION
**Pull request by MarchingCube**
## What does the pull request do?

Improves on current `VisualExtensions` by eliminating LINQ usage and adding a few new functions.

## What is the current behavior?

Only `IEnumerable` extension methods are available which leads to poor performance and many wasteful memory allocations.

## What is the updated/expected behavior with this PR?

100% decrease in allocations when finding common visual ancestor (used by hit testing, etc.)

|                     Method |       Mean |     Error |    StdDev | Ratio |       Gen 0 | Gen 1 | Gen 2 |    Allocated |
|--------------------------- |-----------:|----------:|----------:|------:|------------:|------:|------:|-------------:|
| CommonAncestor_IEnumerable | 3,530.3 ms | 14.935 ms | 13.970 ms |  1.00 | 617000.0000 |     - |     - | 2582357104 B |
|  CommonAncestor_Optimized |   387.7 ms |  2.426 ms |  2.026 ms |  0.11 |           - |     - |     - |            - |

Results from benchmarks included in this PR:

|                       Method |         Mean |         Error |        StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------- |-------------:|--------------:|--------------:|--------:|------:|------:|----------:|
|      FindAncestorOfType_Linq |     695.0 us |     3.6113 us |     3.0156 us | 83.0078 |     - |     - |  348640 B |
| FindAncestorOfType_Optimized |     184.0 us |     0.8506 us |     0.7956 us |       - |     - |     - |         - |
|     FindCommonVisualAncestor | 392,541.6 us | 1,693.0954 us | 1,500.8850 us |       - |     - |     - |         - |

## How was the solution implemented (if it's not obvious)?

After rethinking https://github.com/AvaloniaUI/Avalonia/pull/3247 I've decided to just modify current extension methods and drop custom visitors. It is simpler to just write that code inline and less confusing for future maintainers.

I did change few random places since I wanted to get a feel how we are using visual trees and which parts can be optimized.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

Replaces https://github.com/AvaloniaUI/Avalonia/pull/3247.

### CODE REVIEW

**Comment by jmacato** [comment]
Holy smoking bacon, that perf is :fire:

File: src/Avalonia.Visuals/VisualTree/VisualExtensions.cs [hunk 1]
```diff
@@ -69,6 +151,59 @@ public static IEnumerable<IVisual> GetVisualAncestors(this IVisual visual)
...
+            }
+
+            if (includeSelf && visual is T result)
+            {
+                return result;
+            }
+
+            FindDescendantOfTypeCore<T>(visual);
+
+            return null;
```
Comment (jp2masa) [inline src/Avalonia.Visuals/VisualTree/VisualExtensions.cs]: Isn't this supposed to `return FindDescendantOfTypeCore<T>(visual);`?

File: src/Avalonia.Visuals/VisualTree/VisualExtensions.cs [hunk 2]
```diff
@@ -47,8 +73,64 @@ public static IVisual FindCommonVisualAncestor(this IVisual visual, IVisual targ
...
+                }
+
+                visual = visual.VisualParent;
+                target = target.VisualParent;
+
+                if (visual == null || target == null)
+                {
+                    return null;
+                }
+            }
```
Comment (jp2masa) [inline src/Avalonia.Visuals/VisualTree/VisualExtensions.cs]: Maybe it's clearer and safer to use `while(visual != null && target != null)`?

File: src/Avalonia.Visuals/VisualTree/VisualExtensions.cs (hunk 1 above)
Comment (MarchingCube) [reply src/Avalonia.Visuals/VisualTree/VisualExtensions.cs]: It is wrong indeed. I probably broke it when moving it around.

File: src/Avalonia.Layout/Layoutable.cs [hunk 3]
```diff
@@ -693,14 +692,37 @@ protected virtual Size ArrangeOverride(Size finalSize)
             return finalSize;
         }
 
-        /// <inheritdoc/>
-        protected override sealed void OnVisualParentChanged(IVisual oldParent, IVisual newParent)
+        /// <summary>
+        /// Invalidates measure for this instance and all visual children.
+        /// </summary>
+        protected void InvalidateSelfAndDescendantsMeasure()
```
Comment (grokys) [inline src/Avalonia.Layout/Layoutable.cs]: Hmm, making this a protected method makes it an API, which feel kinda wrong? It's only for use in very particular scenarios, and shouldn't really be used in any old derived control.

I'm not sure what the best alternative is though, given that it's needed in 2 different Avalonia assemblies.

Comment (kekekeks) [reply src/Avalonia.Layout/Layoutable.cs]: It might be time to start generating reference assemblies and strip everything with `[PrivateAPI]` attribute.

Comment (MarchingCube) [reply src/Avalonia.Layout/Layoutable.cs]: Moved it to `LayoutHelper` for now.

**Review by grokys — APPROVED** [review]

**Pull request merged**